File: huffman/src/python/min_priority_queue.py
import heapq
from typing import List
from node import Node

class Min_Priority_Queue:
    def __init__(self, node_list: List[Node] = None):
        if node_list is not None:
            self._heap = [(node.frequency, not node.combined, node) for node in node_list]
            heapq.heapify(self._heap)
        else:
            self._heap = []
    
    def push(self, item: Node):
        """Adds an element to the queue"""
        heapq.heappush(self._heap, (item.frequency, not item.combined, item))
    
    def extract_min(self) -> Node:
        """Removes and returns the element with the minimum priority."""
        if self.is_empty():
            raise IndexError("The queue is empty")
        return heapq.heappop(self._heap)[2]
    
    def peek(self):
        """Returns the element with the minimum priority without removing it."""
        if self.is_empty():
            raise IndexError("The queue is empty")
        return self._heap[0][2]
    
    def is_empty(self):
        """Checks if the queue is empty."""
        return len(self._heap) == 0
    
    def size(self):
        """Returns the number of elements in the queue."""
        return len(self._heap)
    
    def __str__(self):
        """Returns a string representation of the priority queue."""
        return str(sorted(self._heap))

File: huffman/src/c++/classes/min_priority_queue.cpp
#include "min_priority_queue.hpp"

/**
 * Constructor - initializes the heap with given nodes and comparator
 * @param elements Element to insert into minimum priority queue
 * @param comp Compare function to compare minimum priority queue elements
 */
template <typename T, typename Compare>
MinPriorityQueue<T, Compare>::MinPriorityQueue(const vector<T>& elements, Compare comp) : heap(comp, elements) {}

/**
 * Inserts a new element into the priority queue
 */
template <typename T, typename Compare>
void MinPriorityQueue<T, Compare>::insert(const T& value) {
  heap.push(value);
}

/**
 * Removes and returns the minimum element from the queue
 * Returns by value to avoid ownership issues
 */
template <typename T, typename Compare>
T* MinPriorityQueue<T, Compare>::extract_min() {
  if (heap.empty()) {
    throw runtime_error("Priority queue is empty.");
  }
  T* min_value = new T(heap.top());
  heap.pop();
  return min_value;
}

/**
 * Returns a const reference to the minimum element without removing it
 */
template <typename T, typename Compare>
T MinPriorityQueue<T, Compare>::get_min() const {
  if (heap.empty()) {
    throw runtime_error("Priority queue is empty.");
  }
  return heap.top();
}

/**
 * Checks if the priority queue is empty
 */
template <typename T, typename Compare>
bool MinPriorityQueue<T, Compare>::is_empty() const {
  return heap.empty();
}

/**
 * Returns the number of elements in the priority queue
 */
template <typename T, typename Compare>
size_t MinPriorityQueue<T, Compare>::size() const {
  return heap.size();
}

/**
 * Explicit template instantiation for Node type
 */
template class MinPriorityQueue<Node, CompareNodes>;